pkg/clusters/cleanup.go

```diff
@@ -40,6 +44,144 @@ func (c *Cleaner) AddNamespace(namespace *corev1.Namespace) {
+	cmd.Stdout = describeAllOut
+	if err := cmd.Run(); err != nil {
+		return output, err
+	}
+	defer describeAllOut.Close()
+
+	// for each Pod, run kubectl logs
+	pods, err := c.cluster.Client().CoreV1().Pods("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return output, err
+	}
+	logsDir := filepath.Join(output, "pod_logs")
+	err = os.Mkdir(logsDir, 0750) //nolint:gomnd
+	if err != nil {
+		return output, err
+	}
+	failedPods := make(map[string]error)
+	for _, pod := range pods.Items {
+		podLogOut, err := os.Create(filepath.Join(logsDir, fmt.Sprintf("%s_%s", pod.Namespace, pod.Name)))
+		if err != nil {
+			failedPods[fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)] = err
+			continue
+		}
+		cmd := exec.CommandContext(ctx, "kubectl", "--kubeconfig", kubeconfig.Name(), "logs", "-n", pod.Namespace, pod.Name) //nolint:gosec
+		cmd.Stdout = podLogOut
+		if err := cmd.Run(); err != nil {
+			failedPods[fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)] = err
+			continue
+		}
+		defer podLogOut.Close()
+	}
+	if len(failedPods) > 0 {
+		failedPodOut, err := os.Create(filepath.Join(output, "pod_logs_failures.txt"))
+		if err != nil {
+			return output, err
+		}
+		defer failedPodOut.Close()
+		for failed, reason := range failedPods {
+			_, err = failedPodOut.WriteString(fmt.Sprintf("%s: %v\n", failed, reason))
+			if err != nil {
+				return output, err
+			}
+		}
+	}
+
+	// for each Addon, run the addon diagnostic function
+	failedAddons := make(map[string]error)
+	for _, addon := range c.cluster.ListAddons() {
+		diagnostics, err := addon.DumpDiagnostics(ctx, c.cluster)
+		if err != nil {
+			failedAddons[string(addon.Name())] = err
+			continue
+		}
+		if len(diagnostics) > 0 {
+			addonOut := filepath.Join(output, "addons", string(addon.Name()))
+			err = os.MkdirAll(addonOut, 0750) //nolint:gomnd
+			if err != nil {
+				failedAddons[string(addon.Name())] = err
+				continue
+			}
+			for filename, content := range diagnostics {
+				diagOut, err := os.Create(filepath.Join(addonOut, filename))
+				if err != nil {
+					failedAddons[string(addon.Name())] = err
+					continue
+				}
+				defer diagOut.Close()
+				_, err = diagOut.Write(content)
+				if err != nil {
+					failedAddons[string(addon.Name())] = err
+					continue
+				}
+			}
+		}
+	}
+	if len(failedAddons) > 0 {
+		failedAddonOut, err := os.Create(filepath.Join(output, "addon_failures.txt"))
+		if err != nil {
+			return output, err
+		}
+		defer failedAddonOut.Close()
+		for failed, reason := range failedAddons {
+			_, err = failedAddonOut.WriteString(fmt.Sprintf("%s: %v\n", failed, reason))
+			if err != nil {
+				return output, err
+			}
+		}
+	}
+
+	// write the diagnostic metadata
+	metaOut, err := os.Create(filepath.Join(output, "meta.txt"))
+	if err != nil {
+		return output, err
+	}
+	defer metaOut.Close()
+	_, err = metaOut.WriteString(meta)
+	if err != nil {
+		return output, err
+	}
+
+	return output, nil
```

Comment:
So if I'm reading this function right we're stuffing all the diagnostic information into a directory and then providing a string path to that directory. On a local system this would be OK, but what about in CI? Shouldn't we consider having this method build and return a TAR archive of the contents such as to make the process more self-contained and versatile, but also so that it's even more trivial to store the result as a CI artifact and download it?

Reply:
CI works fine, you can write to runner filesystems. If you give upload-artifact multiple files, it automatically bundles them into a single zip, as shown in the Artifacts section at the bottom of https://github.com/rainest/kubernetes-ingress-controller/actions/runs/2741559860